## CP IOT (PARTE 1)

> * RM556219 - Ana Carolina Reis Santana
> * RM558464- Leticia Zago
> * RM555276 - Pedro Henrique Mendonça


In [ ]:
# Importando bibliotecas
import pandas as pd
import numpy as np

from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, mean_squared_error, r2_score
from sklearn.datasets import fetch_california_housing

from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

Dataset: https://archive.ics.uci.edu/dataset/109/wine

### *1. Treinar uma rede neural em Keras para classificar vinhos em 3 classes.*
  - Configuração mínima: 2 camadas ocultas com 32 neurônios cada, função de ativação ReLU.
  - Camada de saída com 3 neurônios, função de ativação Softmax.
  - Função de perda: categorical_crossentropy.
  - Otimizador: Adam

In [ ]:
# Carregando Dados

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data"
column_names = ['Class', 'Alcohol', 'Malic_acid', 'Ash', 'Alcalinity_of_ash', 'Magnesium','Total_phenols', 'Flavanoids', 'Nonflavanoid_phenols', 'Proanthocyanins', 'Color_intensity', 'Hue', 'OD280_OD315_of_diluted_wines', 'Proline']
data = pd.read_csv(url, header=None, names=column_names)


In [ ]:
data.head()

Class  Alcohol  Malic_acid   Ash  Alcalinity_of_ash  Magnesium  \
0      1    14.23        1.71  2.43               15.6        127   
1      1    13.20        1.78  2.14               11.2        100   
2      1    13.16        2.36  2.67               18.6        101   
3      1    14.37        1.95  2.50               16.8        113   
4      1    13.24        2.59  2.87               21.0        118   

   Total_phenols  Flavanoids  Nonflavanoid_phenols  Proanthocyanins  \
0           2.80        3.06                  0.28             2.29   
1           2.65        2.76                  0.26             1.28   
2           2.80        3.24                  0.30             2.81   
3           3.85        3.49                  0.24             2.18   
4           2.80        2.69                  0.39             1.82   

   Color_intensity   Hue  OD280_OD315_of_diluted_wines  Proline  
0             5.64  1.04                          3.92     1065  
1             4.38  1.05                          3.40     1050  
2             5.68  1.03                          3.17     1185  
3             7.80  0.86                          3.45     1480  
4             4.32  1.04                          2.93      735

In [ ]:
data.Class

0      1
1      1
2      1
3      1
4      1
      ..
173    3
174    3
175    3
176    3
177    3
Name: Class, Length: 178, dtype: int64

In [ ]:
# Selecionando as features


X = data.iloc[:, 1:]

In [ ]:
X.shape

(178, 13)

In [ ]:
X.head()

Alcohol  Malic_acid   Ash  Alcalinity_of_ash  Magnesium  Total_phenols  \
0    14.23        1.71  2.43               15.6        127           2.80   
1    13.20        1.78  2.14               11.2        100           2.65   
2    13.16        2.36  2.67               18.6        101           2.80   
3    14.37        1.95  2.50               16.8        113           3.85   
4    13.24        2.59  2.87               21.0        118           2.80   

   Flavanoids  Nonflavanoid_phenols  Proanthocyanins  Color_intensity   Hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   OD280_OD315_of_diluted_wines  Proline  
0                          3.92     1065  
1                          3.40     1050  
2                          3.17     1185  
3                          3.45     1480  
4                          2.93      735

In [ ]:
# Selecionando Dados

y = data['Class']

In [ ]:
y

0      1
1      1
2      1
3      1
4      1
      ..
173    3
174    3
175    3
176    3
177    3
Name: Class, Length: 178, dtype: int64

In [ ]:
#Ajustando para começar do 0

y = y - 1

In [ ]:
# Transformando colunas categoricas em numericas

label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

In [ ]:
# Separando os dados de teste

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
# Dados Normalizados
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)

X_test_scaled = scaler.transform(X_test)

In [ ]:
# Transformando labels categoricas

y_train_categorical = to_categorical(y_train, num_classes=3)

y_test_categorical = to_categorical(y_test, num_classes=3)

In [ ]:
# Modelo

my_nn = Sequential()

In [ ]:
X.shape[1]

13

In [ ]:
# Camada oculta


my_nn.add(Dense(32, input_dim=13, activation='relu'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
# 2 camada

my_nn.add(Dense(32, activation='relu'))

In [ ]:
# Camada de saída

my_nn.add(Dense(3, activation='softmax'))

In [ ]:
my_nn.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 32)             │           448 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 32)             │         1,056 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 3)              │            99 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,603 (6.26 KB)

 Trainable params: 1,603 (6.26 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
# Compilando modelo

my_nn.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [ ]:
# Treinando
history = my_nn.fit(X_train_scaled, y_train_categorical, epochs=100, batch_size=32, validation_data=(X_test_scaled, y_test_categorical), verbose=1)


Epoch 1/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 2s 70ms/step - accuracy: 0.4522 - loss: 1.3020 - val_accuracy: 0.6111 - val_loss: 1.1100
Epoch 2/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.5632 - loss: 1.1518 - val_accuracy: 0.6667 - val_loss: 1.0141
Epoch 3/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 31ms/step - accuracy: 0.5556 - loss: 1.1164 - val_accuracy: 0.6944 - val_loss: 0.9287
Epoch 4/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 35ms/step - accuracy: 0.6628 - loss: 0.9575 - val_accuracy: 0.6944 - val_loss: 0.8489
Epoch 5/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 34ms/step - accuracy: 0.7096 - loss: 0.8509 - val_accuracy: 0.6944 - val_loss: 0.7760
Epoch 6/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step - accuracy: 0.6753 - loss: 0.8193 - val_accuracy: 0.7778 - val_loss: 0.7089
Epoch 7/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 26ms/step - accuracy: 0.7759 - loss: 0.7111 - val_accuracy: 0.8333 - val_loss: 0.6460
Epoch 8/100
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 36ms/step - accuracy: 0.8045 - loss: 0.6468 - val_accuracy: 0.8611 - val_loss:

In [ ]:
# Modelo avaliado

test_loss, test_accuracy = my_nn.evaluate(X_test_scaled, y_test_categorical, verbose=0)


print(f"Acurácia no conjunto de teste: {test_accuracy:.4f}")

Acurácia no conjunto de teste: 1.0000


##### **2. Comparar os resultados com um modelo do scikit-learn (RandomForestClassifier).**

In [ ]:
# Modelo RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators=100, random_state=45)

rf_model.fit(X_train_scaled, y_train)

RandomForestClassifier(random_state=45)

In [ ]:
# Predições Random Forest

y_pred_rf = rf_model.predict(X_test_scaled)

# Avaliar o modelo

accuracy_rf = accuracy_score(y_test, y_pred_rf)
print(f"Acurácia do Random Forest: {accuracy_rf:.4f}")

Acurácia do Random Forest: 1.0000


In [ ]:
# Comparação
print("Comparação de resultados:")
print(f"Acurácia da rede neural em Keras: {test_accuracy:.4f}")
print(f"Acurácia do RandomForestClassifier: {accuracy_rf:.4f}")

Comparação de resultados:
Acurácia da rede neural em Keras: 1.0000
Acurácia do RandomForestClassifier: 1.0000


### **3. Registrar métricas de acurácia e discutir qual modelo teve melhor desempenho.**

Ambos os modelos — a rede neural desenvolvida em Keras e o RandomForestClassifier — obtiveram 100% de acurácia no conjunto de teste. Esse resultado demonstra que, para este dataset, as duas abordagens foram igualmente eficazes, classificando corretamente todas as instâncias avaliadas. Entretanto, é importante destacar que valores perfeitos de acurácia podem indicar overfitting, dependendo da complexidade do modelo e da representatividade dos dados de teste.

### **Regressão**

Dataset: https://archive.ics.uci.edu/ml/datasets/automobile

In [ ]:
# Carregar o conjunto de dados diretamente do link da UCI
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data"

# Nomes das colunas conforme documentação do dataset
feature_names = [
    "symboling", "normalized_losses", "make", "fuel_type", "aspiration", "num_of_doors",
    "body_style", "drive_wheels", "engine_location", "wheel_base", "length", "width",
    "height", "curb_weight", "engine_type", "num_of_cylinders", "engine_size",
    "fuel_system", "bore", "stroke", "compression_ratio", "horsepower", "peak_rpm",
    "city_mpg", "highway_mpg", "price"
]

# Ler o CSV
auto = pd.read_csv(url, names=feature_names)

# Tratar valores ausentes e substituir "?" por NaN
auto = auto.replace("?", pd.NA).dropna()

# Converter colunas numéricas para float
numeric_cols = [
    "normalized_losses", "wheel_base", "length", "width", "height", "curb_weight",
    "engine_size", "bore", "stroke", "compression_ratio", "horsepower", "peak_rpm",
    "city_mpg", "highway_mpg", "price"
]
auto[numeric_cols] = auto[numeric_cols].astype(float)

# Criar X e y no mesmo formato do exemplo original
X_auto = auto.drop(columns=["price"])
y_auto = auto["price"]

print(X_auto.head())
print(y_auto.head())


    symboling  normalized_losses  make fuel_type aspiration num_of_doors  \
3           2              164.0  audi       gas        std         four   
4           2              164.0  audi       gas        std         four   
6           1              158.0  audi       gas        std         four   
8           1              158.0  audi       gas      turbo         four   
10          2              192.0   bmw       gas        std          two   

   body_style drive_wheels engine_location  wheel_base  ...  num_of_cylinders  \
3       sedan          fwd           front        99.8  ...              four   
4       sedan          4wd           front        99.4  ...              five   
6       sedan          fwd           front       105.8  ...              five   
8       sedan          fwd           front       105.8  ...              five   
10      sedan          rwd           front       101.2  ...              four   

    engine_size  fuel_system  bore stroke compression_ra

In [ ]:
X_auto.head()

symboling  normalized_losses  make fuel_type aspiration num_of_doors  \
3           2              164.0  audi       gas        std         four   
4           2              164.0  audi       gas        std         four   
6           1              158.0  audi       gas        std         four   
8           1              158.0  audi       gas      turbo         four   
10          2              192.0   bmw       gas        std          two   

   body_style drive_wheels engine_location  wheel_base  ...  num_of_cylinders  \
3       sedan          fwd           front        99.8  ...              four   
4       sedan          4wd           front        99.4  ...              five   
6       sedan          fwd           front       105.8  ...              five   
8       sedan          fwd           front       105.8  ...              five   
10      sedan          rwd           front       101.2  ...              four   

    engine_size  fuel_system  bore stroke compression_ratio  horsepower  \
3         109.0         mpfi  3.19    3.4              10.0       102.0   
4         136.0         mpfi  3.19    3.4               8.0       115.0   
6         136.0         mpfi  3.19    3.4               8.5       110.0   
8         131.0         mpfi  3.13    3.4               8.3       140.0   
10        108.0         mpfi  3.50    2.8               8.8       101.0   

   peak_rpm  city_mpg  highway_mpg  
3    5500.0      24.0         30.0  
4    5500.0      18.0         22.0  
6    5500.0      19.0         25.0  
8    5500.0      17.0         20.0  
10   5800.0      23.0         29.0  

[5 rows x 25 columns]

In [ ]:
X_auto.shape

(159, 25)

In [ ]:
y_auto.shape

(159,)

In [ ]:
# Informações sobre o target
print("Valor médio dos carros:")
print(f"Mínimo: ${y_auto.min():.2f}")
print(f"Máximo: ${y_auto.max():.2f}")
print(f"Média: ${y_auto.mean():.2f}")

Valor médio dos carros:
Mínimo: $5118.00
Máximo: $35056.00
Média: $11445.73


In [ ]:
# Separação de dados de treino e teste
X_train_auto, X_test_auto, y_train_auto, y_test_auto = train_test_split(X_auto, y_auto, test_size=0.2, random_state=42)

In [ ]:
X_train_auto.shape

(127, 25)

In [ ]:
X_test_auto.shape

(32, 25)

In [ ]:
# Identify categorical and numerical columns
categorical_cols = X_train_auto.select_dtypes(include='object').columns
numerical_cols = X_train_auto.select_dtypes(exclude='object').columns

# Apply one-hot encoding to categorical columns
X_train_auto_encoded = pd.get_dummies(X_train_auto, columns=categorical_cols, dummy_na=False)
X_test_auto_encoded = pd.get_dummies(X_test_auto, columns=categorical_cols, dummy_na=False)

# Align columns after one-hot encoding - important for consistent feature sets
X_train_auto_encoded, X_test_auto_encoded = X_train_auto_encoded.align(X_test_auto_encoded, join='inner', axis=1, fill_value=0)


# Normalization of numerical data
scaler_auto = StandardScaler()
X_train_auto_scaled = scaler_auto.fit_transform(X_train_auto_encoded)
X_test_auto_scaled = scaler_auto.transform(X_test_auto_encoded)

In [ ]:
# Construção do modelo
my_nn_regression = Sequential()

In [ ]:
# 1ª camada oculta: 64 neurônios
my_nn_regression.add(Dense(64, input_dim=X_train_auto_scaled.shape[1], activation='relu'))

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:93: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [ ]:
# 2ª camada oculta: 32 neurônios
my_nn_regression.add(Dense(32, activation='relu'))

In [ ]:
# Camada de saída com 1 neurônio, função de ativação Linear
my_nn_regression.add(Dense(1, activation='linear'))

In [ ]:
my_nn_regression.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 64)             │         3,264 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 5,377 (21.00 KB)

 Trainable params: 5,377 (21.00 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
# Compilar o modelo
my_nn_regression.compile(optimizer='adam', loss='mse', metrics=['mae'])

In [ ]:
# Treinamento do modelo
history_regression = my_nn_regression.fit(X_train_auto_scaled, y_train_auto, epochs=100, batch_size=32,
                                          validation_data=(X_test_auto_scaled, y_test_auto), verbose=1)

Epoch 1/100
4/4 ━━━━━━━━━━━━━━━━━━━━ 2s 114ms/step - loss: 163376416.0000 - mae: 11432.0605 - val_loss: 132734080.0000 - val_mae: 10720.9619
Epoch 2/100
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 27ms/step - loss: 170969632.0000 - mae: 11428.2510 - val_loss: 132723576.0000 - val_mae: 10720.5156
Epoch 3/100
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 22ms/step - loss: 177666336.0000 - mae: 11661.0488 - val_loss: 132713448.0000 - val_mae: 10720.0869
Epoch 4/100
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 23ms/step - loss: 168262032.0000 - mae: 11578.8379 - val_loss: 132704160.0000 - val_mae: 10719.6895
Epoch 5/100
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 25ms/step - loss: 177437776.0000 - mae: 11750.3691 - val_loss: 132695216.0000 - val_mae: 10719.3086
Epoch 6/100
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 22ms/step - loss: 166211456.0000 - mae: 11464.5898 - val_loss: 132686320.0000 - val_mae: 10718.9287
Epoch 7/100
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step - loss: 159275584.0000 - mae: 11219.9717 - val_loss: 132677072.0000 - val_mae: 10718.5400
Epoch 8/100
4/4 ━━━

In [ ]:
# Avaliação da rede neural
test_loss_regression, test_mae_regression = my_nn_regression.evaluate(X_test_auto_scaled, y_test_auto, verbose=0)
print(f"MSE da rede neural: {test_loss_regression:.4f}")
print(f"MAE da rede neural: {test_mae_regression:.4f}")

MSE da rede neural: 91073936.0000
MAE da rede neural: 9068.5332


In [ ]:
# Predições da rede neural
y_pred_nn_regression = my_nn_regression.predict(X_test_auto_scaled)
r2_nn = r2_score(y_test_auto, y_pred_nn_regression)
print(f"R² da rede neural: {r2_nn:.4f}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 68ms/step
R² da rede neural: -4.1181


##### **2. Comparar os resultados com um modelo do scikit-learn (LinearRegression).**

In [ ]:
# Treinando modelo LinearRegression
lr_regression = LinearRegression()
lr_regression.fit(X_train_auto_scaled, y_train_auto)

LinearRegression()

In [ ]:
# Predições LinearRegression
y_pred_lr_regression = lr_regression.predict(X_test_auto_scaled)
mse_lr = mean_squared_error(y_test_auto, y_pred_lr_regression)
r2_lr = r2_score(y_test_auto, y_pred_lr_regression)
print(f"MSE da LinearRegression: {mse_lr:.4f}")
print(f"R² da LinearRegression: {r2_lr:.4f}")

MSE da LinearRegression: 5704165.3940
R² da LinearRegression: 0.6794


In [ ]:
# RMSE da rede neural
rmse_nn = np.sqrt(test_loss_regression)
print(f"RMSE da Rede Neural: {rmse_nn:.4f}")

RMSE da Rede Neural: 9543.2665


In [ ]:
# RMSE da LinearRegression
rmse_lr = np.sqrt(mse_lr)
print(f"RMSE da Linear Regression: {rmse_lr:.4f}")

RMSE da Linear Regression: 2388.3395


In [ ]:
# Comparar os resultados
print("Comparação de resultados:")
print(f"Acurácia da rede neural MSE: {test_loss_regression:.4f}, RMSE: {rmse_nn:.4f}, R²: {r2_nn:.4f}")
print(f"Acurácia do LinearRegression: {mse_lr:.4f}, RMSE: {rmse_lr:.4f}, R²: {r2_lr:.4f}")

Comparação de resultados:
Acurácia da rede neural MSE: 91073936.0000, RMSE: 9543.2665, R²: -4.1181
Acurácia do LinearRegression: 5704165.3940, RMSE: 2388.3395, R²: 0.6794


##### **3. Registrar métricas de erro (RMSE ou MAE) e discutir qual modelo teve melhor desempenho.**

**Métricas de erro:**
A análise das métricas mostra que o modelo de Regressão Linear apresentou desempenho significativamente superior ao da rede neural neste conjunto de dados.

A Regressão Linear alcançou um erro médio (RMSE) muito menor e um R² positivo (0,68), indicando que conseguiu explicar cerca de 68% da variabilidade dos preços.

Por outro lado, a rede neural exibiu erros extremamente elevados e um R² negativo, o que sugere que ela não conseguiu generalizar bem e teve desempenho pior que uma simples média dos valores.

Com base nessas métricas, a rede neural em Keras demonstrou ter um desempenho superior na tarefa de prever o valor médio das casas neste dataset, com erros de previsão menores e uma maior capacidade de explicar a variância nos dados.